cnxdb/migrations/20170911201035_transform_cnxml_to_html.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+import logging
+import time
+from datetime import timedelta
+
+import rhaptos.cnxmlutils
+from dbmigrator import deferred, logger
+
+
+def _batcher(seq, size):
+    for pos in range(0, len(seq), size):
+         yield seq[pos:pos + size]
+
+
+def should_run(cursor, limit='LIMIT 1'):
+    version = rhaptos.cnxmlutils.__version__
+    version_text = '%data-cnxml-to-html-ver="{}"%'.format(version)
+    logger.info('Looking for {}'.format(version_text))
+    cursor.execute("""\
+SELECT files.fileid, module_ident
+    FROM module_files
+    NATURAL JOIN files
+    WHERE filename = 'index.cnxml.html'
+      AND convert_from(file, 'utf-8') NOT LIKE %s {}""".format(limit),
+                   (version_text,))
+    return cursor.fetchall()
+
+
+@deferred
+def up(cursor):
+    """Transform all index.cnxml to index.cnxml.html"""
+    # Get all the index.cnxml.html that does not have rhaptos.cnxmlutils
+    # current version
+    to_transform = should_run(cursor, limit='')
+    num_todo = len(to_transform)
+
+    batch_size = 10
+    logger.info('Batch size: {}'.format(batch_size))
+
+    start = time.time()
+    guesstimate = 1 * num_todo
+    guess_complete = guesstimate + start
+    logger.info('Completion guess: '
+                '"{}" ({})'.format(time.ctime(guess_complete),
+                                   timedelta(0, guesstimate)))
+
+    cursor.execute('DROP TABLE IF EXISTS datamigrations.index_cnxml_html')
+
+    # Store module_files for modules we are going to update
+    cursor.execute("""\
+CREATE TABLE datamigrations.index_cnxml_html AS
+    SELECT * FROM module_files
+    WHERE module_ident IN %s
+      AND filename = 'index.cnxml.html'""",
+                   (tuple([i[1] for i in to_transform]),))
+
+    num_complete = 0
+    for batch in _batcher(to_transform, batch_size):
+        module_idents = tuple([i[1] for i in batch])
+        logger.info('Transform module_idents {}'.format(module_idents))
+
+        for fileid, module_ident in batch:
+            cursor.execute("""\
+INSERT INTO files
+    (file, media_type)
+    SELECT convert_to(html_content(%s), 'utf-8'), 'text/xml'
+RETURNING fileid""", (module_ident,))
```

Comment:
It seems there might be some index.cnxml that are shared? :thinking:

```
Traceback (most recent call last):
  File "/var/cnx/venvs/publishing/bin/dbmigrator", line 11, in <module>
    sys.exit(main())
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/cli.py", line 104, in main
    return args['cmmd'](**args)
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/utils.py", line 145, in wrapper
    return func(cursor, *args, **kwargs)
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/commands/migrate.py", line 32, in cli_command
    run_deferred)
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/utils.py", line 227, in compare_schema
    callback(*args, **kwargs)
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/utils.py", line 257, in run_migration
    migration.up(cursor)
  File "cnxdb/migrations/20170911201035_transform_cnxml_to_html.py", line 67, in up
    RETURNING fileid""", (module_ident,))
psycopg2.IntegrityError: duplicate key value violates unique constraint "files_sha1_key"
DETAIL:  Key (sha1)=(45c6c6d1a97925ed5ebcc57711a2119477d7bf75) already exists.
```

Reply:
Looks like there are...

```
cnxarchive=# select count(*), fileid, array_agg(module_ident) from module_files where filename = 'index.cnxml' group by fileid having count(*) > 1;
 count | fileid |                  array_agg                  
-------+--------+---------------------------------------------
     2 | 372499 | {22292,8905}
     2 | 374610 | {17325,17324}
     2 | 377685 | {31800,31768}
     2 | 377706 | {31806,31807}
     2 | 377709 | {31811,31809}
     7 | 377716 | {31818,31820,31817,31824,31823,31822,31821}
(6 rows)
```

Reply:
Changing `should_run` to use `max(module_ident)` and `GROUP BY files.fileid` is probably going to fix the problem (correctly?).

Reply:
Production does have 669 of these as well. All of them seem to be shared between adjacent versions of the same module - _i.e._ the author republished w/ no changes (which perhaps we should detect). Archive does data deduplication at the file level, so when these are stored, the same fileid is returned. No error.